file.go

```diff
@@ -10,14 +14,104 @@ import (
 var (
 	_ fs.File     = (*file)(nil)
 	_ fs.FileInfo = (*fileInfo)(nil)
+	_ io.Seeker   = (*file)(nil)
 )
 
 type file struct {
-	io.ReadCloser
-	stat func() (fs.FileInfo, error)
+	cl     s3iface.S3API
+	bucket string
+	name   string
+
+	realReader      io.ReadCloser
+	stat            func() (fs.FileInfo, error)
+	currentPosition int64
+}
+
+func NewFile(cl s3iface.S3API, bucket string, name string) (fs.File, error) {
+	out, err := cl.GetObject(&s3.GetObjectInput{
+		Key:    &name,
+		Bucket: &bucket,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	statFunc := func() (fs.FileInfo, error) {
+		return stat(cl, bucket, name)
+	}
+
+	if out.ContentLength != nil && out.LastModified != nil {
+		// if we got all the information from GetObjectOutput
+		// then we can cache fileinfo instead of making
+		// another call in case Stat is called.
+		statFunc = func() (fs.FileInfo, error) {
+			return &fileInfo{
+				name:    path.Base(name),
+				size:    *out.ContentLength,
+				modTime: *out.LastModified,
+			}, nil
+		}
+	}
+
+	return &file{
+		cl:         cl,
+		bucket:     bucket,
+		name:       name,
+		realReader: out.Body,
+		stat:       statFunc,
+	}, nil
+}
+
+func (f *file) Read(p []byte) (int, error) {
+	n, err := f.realReader.Read(p)
+	f.currentPosition += int64(n)
+	return n, err
+}
+
+func (f *file) Close() error {
+	return f.realReader.Close()
+}
+
+func (f *file) Seek(offset int64, whence int) (int64, error) {
+	newPosition := f.currentPosition
+	switch whence {
+	case io.SeekStart:
+		newPosition = offset
+	case io.SeekCurrent:
+		newPosition += offset
+	case io.SeekEnd:
+		stat, err := f.Stat()
+		if err != nil {
+			return 0, err
+		}
+		newPosition = stat.Size() + offset
+	default:
+		return 0, fmt.Errorf("unknown 'whence': %d", whence)
+	}
+	// If the position has not moved, there is no need to make a new query
+	if f.currentPosition == newPosition {
+		return newPosition, nil
+	}
+
+	rawObject, err := f.cl.GetObject(
+		&s3.GetObjectInput{
+			Bucket: aws.String(f.bucket),
+			Key:    aws.String(f.name),
+			Range:  aws.String(fmt.Sprintf("bytes=%d-", newPosition)),
+		})
+
+	if err != nil {
+		return f.currentPosition, err
+	}
+
+	f.currentPosition = newPosition
+	f.realReader = rawObject.Body
+
+	return f.currentPosition, nil
 }
 
-func (f file) Stat() (fs.FileInfo, error) { return f.stat() }
+func (f *file) Stat() (fs.FileInfo, error) { return f.stat() }
```

Comment:
why change this?

Reply:
You are right. That's not needed. reverted